### ISSUE TITLE
Schemas containing reference to another schemas cause confluent lib fail with ccompat/v7 and No artifact with ID error code: 40401

### ISSUE THREAD

**Issue by MilanMasek** [Apicurio/apicurio-registry#5133]
## Description
We suspect an issue with the apps using confluent library, targeting ccompat compatibility api and retrieving messages containing schema with references to another schemas.

**Registry
Version**: 2.5.11 (latest in quay.io for the operator)
**Persistence type**: "kafkasql"

### Environment

What is the environment in which you are running Apicurio Registry?
OCP, Server Version: 4.14.20
Kubernetes Version: v1.27.11+ec42b99
How are you interacting with Apicurio Registry? [kafka-avro-console-consumer](https://github.com/confluentinc/schema-registry/blob/master/bin/kafka-avro-console-consumer) or [kafka-ui](https://github.com/provectus/kafka-ui). 
When our custom apps are retrieving schema ID directly from the message body, there is no issue.

### Steps to Reproduce

P.S. Following scenario succeeds when schema has no references to other schemas.

Preparation step, because we use TLS secured Kafka
```
$ cat /home/masmil/tmp/kafka-eps.t.client.properties
security.protocol=SSL
ssl.truststore.location=/tmp/kafka-eps.eps-shared.t.kafka-eps-admin.keystore.jks
ssl.truststore.password=G7NNnF8dU5vbs1B1l21G4EEa3lAtdmL1
ssl.keystore.location=/tmp/kafka-eps.eps-shared.t.kafka-eps-admin.keystore.jks
ssl.keystore.password=****
ssl.endpoint.identification.algorithm=
```

When retrieving message by kafka-avro-console-consumer, it fails as follows with 

```
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: No artifact with ID 'cz.ifortuna.eps.offer.schema.command.CommandTimestamps' in group 'null' was found.; error code: 40401
```

The error (reported ID/name of the schema) is little bit misleading, because it is referenced schema from "parent" schema.

Full log of the kafka-avro-console-consumer

```
$ podman run -v /tmp/kafka-eps.eps-shared.t.kafka-eps-admin.keystore.jks:/tmp/kafka-eps.eps-shared.t.kafka-eps-admin.keystore.jks -v /home/masmil/tmp/kafka-eps.t.client.properties:/tmp/kafka-eps.t.client.properties confluentinc/cp-schema-registry kafka-avro-console-consumer \
--bootstrap-server kafka-eps.t.dc1.cz.ipa.ifortuna.cz:9190 \
--property schema.registry.url=http://kafka-eps.t.dc1.cz.ipa.ifortuna.cz:9180/apis/ccompat/v7 \
--topic beter-offer-commands \
--consumer.config /tmp/kafka-eps.t.client.properties \
--property print.key=true \
--property print.schema.ids=true \
--property print.value=true \
--value-deserializer io.confluent.kafka.serializers.KafkaAvroDeserializer \
--key-deserializer org.apache.kafka.common.serialization.StringDeserializer \
--partition 7 \
--offset 94896 \
--max-messages 1
[2024-09-06 06:34:07,098] INFO KafkaAvroDeserializerConfig values:
    auto.register.schemas = true
    avro.reflection.allow.null = false
    avro.use.logical.type.converters = false
    basic.auth.credentials.source = URL
    basic.auth.user.info = [hidden]
    bearer.auth.cache.expiry.buffer.seconds = 300
    bearer.auth.client.id = null
    bearer.auth.client.secret = null
    bearer.auth.credentials.source = STATIC_TOKEN
    bearer.auth.custom.provider.class = null
    bearer.auth.identity.pool.id = null
    bearer.auth.issuer.endpoint.url = null
    bearer.auth.logical.cluster = null
    bearer.auth.scope = null
    bearer.auth.scope.claim.name = scope
    bearer.auth.sub.claim.name = sub
    bearer.auth.token = [hidden]
    context.name.strategy = class io.confluent.kafka.serializers.context.NullContextNameStrategy
    http.connect.timeout.ms = 60000
    http.read.timeout.ms = 60000
    id.compatibility.strict = true
    key.subject.name.strategy = class io.confluent.kafka.serializers.subject.TopicNameStrategy
    latest.cache.size = 1000
    latest.cache.ttl.sec = -1
    latest.compatibility.strict = true
    max.schemas.per.subject = 1000
    normalize.schemas = false
    proxy.host =
    proxy.port = -1
    rule.actions = []
    rule.executors = []
    rule.service.loader.enable = true
    schema.format = null
    schema.reflection = false
    schema.registry.basic.auth.user.info = [hidden]
    schema.registry.ssl.cipher.suites = null
    schema.registry.ssl.enabled.protocols = [TLSv1.2, TLSv1.3]
    schema.registry.ssl.endpoint.identification.algorithm = https
    schema.registry.ssl.engine.factory.class = null
    schema.registry.ssl.key.password = null
    schema.registry.ssl.keymanager.algorithm = SunX509
    schema.registry.ssl.keystore.certificate.chain = null
    schema.registry.ssl.keystore.key = null
    schema.registry.ssl.keystore.location = null
    schema.registry.ssl.keystore.password = null
    schema.registry.ssl.keystore.type = JKS
    schema.registry.ssl.protocol = TLSv1.3
    schema.registry.ssl.provider = null
    schema.registry.ssl.secure.random.implementation = null
    schema.registry.ssl.trustmanager.algorithm = PKIX
    schema.registry.ssl.truststore.certificates = null
    schema.registry.ssl.truststore.location = null
    schema.registry.ssl.truststore.password = null
    schema.registry.ssl.truststore.type = JKS
    schema.registry.url = [http://kafka-eps.t.dc1.cz.ipa.ifortuna.cz:9180/apis/ccompat/v7]
    specific.avro.key.type = null
    specific.avro.reader = false
    specific.avro.value.type = null
    use.latest.version = false
    use.latest.with.metadata = null
    use.schema.id = -1
    value.subject.name.strategy = class io.confluent.kafka.serializers.subject.TopicNameStrategy
 (io.confluent.kafka.serializers.KafkaAvroDeserializerConfig:375)
bt:mkt:6902361a-4d25-436c-a333-b23b30d3cf0e:[35,5,6,0,[24-26]]  [2024-09-06 06:34:10,785] ERROR Could not parse Avro schema (io.confluent.kafka.schemaregistry.avro.AvroSchemaProvider:59)
java.lang.RuntimeException: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: No artifact with ID 'cz.ifortuna.eps.offer.schema.command.CommandTimestamps' in group 'null' was found.; error code: 40401
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getByVersion(CachedSchemaRegistryClient.java:503)
    at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences(AbstractSchemaProvider.java:66)
    at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences(AbstractSchemaProvider.java:52)
    at io.confluent.kafka.schemaregistry.avro.AvroSchemaProvider.parseSchemaOrElseThrow(AvroSchemaProvider.java:52)
    at io.confluent.kafka.schemaregistry.SchemaProvider.parseSchema(SchemaProvider.java:75)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.parseSchema(CachedSchemaRegistryClient.java:301)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaByIdFromRegistry(CachedSchemaRegistryClient.java:340)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaBySubjectAndId(CachedSchemaRegistryClient.java:464)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer$DeserializationContext.schemaFromRegistry(AbstractKafkaAvroDeserializer.java:401)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:191)
    at io.confluent.kafka.formatter.AvroMessageFormatter$AvroMessageDeserializer.deserialize(AvroMessageFormatter.java:134)
    at io.confluent.kafka.formatter.AvroMessageFormatter.writeTo(AvroMessageFormatter.java:89)
    at io.confluent.kafka.formatter.SchemaMessageFormatter.writeTo(SchemaMessageFormatter.java:267)
    at kafka.tools.ConsoleConsumer$.process(ConsoleConsumer.scala:117)
    at kafka.tools.ConsoleConsumer$.run(ConsoleConsumer.scala:77)
    at kafka.tools.ConsoleConsumer$.main(ConsoleConsumer.scala:54)
    at kafka.tools.ConsoleConsumer.main(ConsoleConsumer.scala)
Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: No artifact with ID 'cz.ifortuna.eps.offer.schema.command.CommandTimestamps' in group 'null' was found.; error code: 40401
    at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest(RestService.java:340)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest(RestService.java:413)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion(RestService.java:1019)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion(RestService.java:1006)
    at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion(RestService.java:994)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaByVersionFromRegistry(CachedSchemaRegistryClient.java:545)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaByVersion(CachedSchemaRegistryClient.java:524)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getByVersion(CachedSchemaRegistryClient.java:501)
    ... 16 more
Processed a total of 1 messages
[2024-09-06 06:34:11,090] ERROR Unknown error when running consumer:  (kafka.tools.ConsoleConsumer$:45)
org.apache.kafka.common.errors.SerializationException: Error retrieving Avro value schema for id 1798
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer$DeserializationContext.schemaFromRegistry(AbstractKafkaAvroDeserializer.java:410)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:191)
    at io.confluent.kafka.formatter.AvroMessageFormatter$AvroMessageDeserializer.deserialize(AvroMessageFormatter.java:134)
    at io.confluent.kafka.formatter.AvroMessageFormatter.writeTo(AvroMessageFormatter.java:89)
    at io.confluent.kafka.formatter.SchemaMessageFormatter.writeTo(SchemaMessageFormatter.java:267)
    at kafka.tools.ConsoleConsumer$.process(ConsoleConsumer.scala:117)
    at kafka.tools.ConsoleConsumer$.run(ConsoleConsumer.scala:77)
    at kafka.tools.ConsoleConsumer$.main(ConsoleConsumer.scala:54)
    at kafka.tools.ConsoleConsumer.main(ConsoleConsumer.scala)
Caused by: java.io.IOException: Invalid schema {"type":"record","name":"SettleMarket","namespace":"cz.ifortuna.eps.offer.schema.command.market","doc":"Class representation of the command which settles a market","fields":[{"name":"timestamps","type":"cz.ifortuna.eps.offer.schema.command.CommandTimestamps","doc":"Timestamps of command life-cycle transitions"},{"name":"entityId","type":"cz.ifortuna.eps.offer.schema.URID","doc":"Provider identification of the entity this command belongs to"},{"name":"entityExtIds","type":["null",{"type":"array","items":"cz.ifortuna.eps.offer.schema.URID"}],"doc":"External identifications of the entity this command belongs to","default":null},{"name":"entityVersion","type":["null",{"type":"string","avro.java.string":"String"}],"doc":"Calculated version of the entity state this command was created from (only where it makes sense)","default":null},{"name":"status","type":"cz.ifortuna.eps.offer.schema.entities.market.OfferStatus","doc":"The offer status of this market"},{"name":"resolutionTime","type":{"type":"long","logicalType":"timestamp-millis"},"doc":"Time when the market was resulted (milliseconds since EPOCH UTC)"},{"name":"options","type":{"type":"array","items":{"type":"record","name":"SettleOption","doc":"Class representation of a market option to be settled","fields":[{"name":"entityId","type":"cz.ifortuna.eps.offer.schema.URID","doc":"Internal unique identification of this option"},{"name":"entityExtIds","type":["null",{"type":"array","items":"cz.ifortuna.eps.offer.schema.URID"}],"doc":"External identifications of this option","default":null},{"name":"status","type":"cz.ifortuna.eps.offer.schema.entities.market.OfferStatus","doc":"The offer status of this option"},{"name":"result","type":{"type":"record","name":"OptionResult","namespace":"cz.ifortuna.eps.offer.schema.entities.market","doc":"Class representation of an option result","fields":[{"name":"resultType","type":{"type":"enum","name":"OptionResultType","doc":"Enumeration of possible option result types","symbols":["Winning","Losing","Voided"]},"doc":"Value of the option's result type"},{"name":"resolutionTime","type":{"type":"long","logicalType":"timestamp-millis"},"doc":"Time when the option was resulted (milliseconds since EPOCH UTC)"}]},"doc":"The resolution of this option"},{"name":"divisionFactor","type":["null","int"],"doc":"The number the option odds should be divided by during payout","default":null}]}},"doc":"Collection of market options settlements"}]} with refs [{name='cz.ifortuna.eps.offer.schema.command.CommandTimestamps', subject='cz.ifortuna.eps.offer.schema.command.CommandTimestamps', version=1}, {name='cz.ifortuna.eps.offer.schema.URID', subject='cz.ifortuna.eps.offer.schema.URID', version=1}, {name='cz.ifortuna.eps.offer.schema.entities.market.OfferStatus', subject='cz.ifortuna.eps.offer.schema.entities.market.OfferStatus', version=1}] of type AVRO
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.lambda$getSchemaByIdFromRegistry$6(CachedSchemaRegistryClient.java:343)
    at java.base/java.util.Optional.orElseThrow(Optional.java:403)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaByIdFromRegistry(CachedSchemaRegistryClient.java:341)
    at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaBySubjectAndId(CachedSchemaRegistryClient.java:464)
    at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer$DeserializationContext.schemaFromRegistry(AbstractKafkaAvroDeserializer.java:401)
    ... 8 more
```

### Expected vs Actual Behaviour

Expected behavior is that compatibility api returns schema including referenced schemas.

### Logs

Success case, when schema has no referenced schemas.

```
$ podman run -v /tmp/kafka-eps.eps-shared.t.kafka-eps-admin.keystore.jks:/tmp/kafka-eps.eps-shared.t.kafka-eps-admin.keystore.jks -v /home/masmil/tmp/kafka-eps.t.client.properties:/tmp/kafka-eps.t.client.properties confluentinc/cp-schema-registry kafka-avro-console-consumer \
--bootstrap-server kafka-eps.t.dc1.cz.ipa.ifortuna.cz:9190 \
--property schema.registry.url=http://kafka-eps.t.dc1.cz.ipa.ifortuna.cz:9180/apis/ccompat/v7 \
--topic beter-feed \
--consumer.config /tmp/kafka-eps.t.client.properties \
--property print.key=true \
--property print.schema.ids=true \
--property print.value=true \
--value-deserializer io.confluent.kafka.serializers.KafkaAvroDeserializer \
--key-deserializer org.apache.kafka.common.serialization.StringDeserializer \
--partition 5 \
--offset 252862 \
--max-messages 1
[2024-09-06 06:29:37,239] INFO KafkaAvroDeserializerConfig values:
  auto.register.schemas = true
  avro.reflection.allow.null = false
  avro.use.logical.type.converters = false
  basic.auth.credentials.source = URL
  basic.auth.user.info = [hidden]
  bearer.auth.cache.expiry.buffer.seconds = 300
  bearer.auth.client.id = null
  bearer.auth.client.secret = null
  bearer.auth.credentials.source = STATIC_TOKEN
  bearer.auth.custom.provider.class = null
  bearer.auth.identity.pool.id = null
  bearer.auth.issuer.endpoint.url = null
  bearer.auth.logical.cluster = null
  bearer.auth.scope = null
  bearer.auth.scope.claim.name = scope
  bearer.auth.sub.claim.name = sub
  bearer.auth.token = [hidden]
  context.name.strategy = class io.confluent.kafka.serializers.context.NullContextNameStrategy
  http.connect.timeout.ms = 60000
  http.read.timeout.ms = 60000
  id.compatibility.strict = true
  key.subject.name.strategy = class io.confluent.kafka.serializers.subject.TopicNameStrategy
  latest.cache.size = 1000
  latest.cache.ttl.sec = -1
  latest.compatibility.strict = true
  max.schemas.per.subject = 1000
  normalize.schemas = false
  proxy.host =
  proxy.port = -1
  rule.actions = []
  rule.executors = []
  rule.service.loader.enable = true
  schema.format = null
  schema.reflection = false
  schema.registry.basic.auth.user.info = [hidden]
  schema.registry.ssl.cipher.suites = null
  schema.registry.ssl.enabled.protocols = [TLSv1.2, TLSv1.3]
  schema.registry.ssl.endpoint.identification.algorithm = https
  schema.registry.ssl.engine.factory.class = null
  schema.registry.ssl.key.password = null
  schema.registry.ssl.keymanager.algorithm = SunX509
  schema.registry.ssl.keystore.certificate.chain = null
  schema.registry.ssl.keystore.key = null
  schema.registry.ssl.keystore.location = null
  schema.registry.ssl.keystore.password = null
  schema.registry.ssl.keystore.type = JKS
  schema.registry.ssl.protocol = TLSv1.3
  schema.registry.ssl.provider = null
  schema.registry.ssl.secure.random.implementation = null
  schema.registry.ssl.trustmanager.algorithm = PKIX
  schema.registry.ssl.truststore.certificates = null
  schema.registry.ssl.truststore.location = null
  schema.registry.ssl.truststore.password = null
  schema.registry.ssl.truststore.type = JKS
  schema.registry.url = [http://kafka-eps.t.dc1.cz.ipa.ifortuna.cz:9180/apis/ccompat/v7]
  specific.avro.key.type = null
  specific.avro.reader = false
  specific.avro.value.type = null
  use.latest.version = false
  use.latest.with.metadata = null
  use.schema.id = -1
  value.subject.name.strategy = class io.confluent.kafka.serializers.subject.TopicNameStrategy
(io.confluent.kafka.serializers.KafkaAvroDeserializerConfig:375)
b8913881-4326-4cb4-8813-76bcd8743131_[10,5,25,0,[13.5]]  {"id":"b8913881-4326-4cb4-8813-76bcd8743131","sportId":7,"tradingStatus":1,"lineType":1,"willBeLive":true,"willBePrematch":true,"markets":[{"id":"[10,5,25,0,[13.5]]","interval":25,"resultType":5,"marketType":10,"marketValue":"13.5","subinterval":0,"outcomes":[{"id":"[5,[]]","outcomeType":5,"outcomeValue":"","price":1.77,"status":2,"outcomeResult":0,"prices":{"decimal":1.77,"fractional":null}},{"id":"[4,[]]","outcomeType":4,"outcomeValue":"","price":1.93,"status":2,"outcomeResult":0,"prices":{"decimal":1.93,"fractional":null}}]}],"timestamp":1725541057583000,"msgType":1,"offset":1704238944568000}  1777
Processed a total of 1 messages
```

**Comment by carlesarnal** [comment]
Hello @MilanMasek,

The groups feature is meant to be used by our main API. What we did was, in the case of the ccompat API, default all operations to the default group, as you correctly discovered. That said, there are a couple of options:

- Pass the group you want to use in a header -> https://github.com/Apicurio/apicurio-registry/blob/main/common/src/main/java/io/apicurio/registry/rest/Headers.java#L9
- Use the option introduced in https://github.com/Apicurio/apicurio-registry/pull/5017 to fully operate with groups.
- Of course, just use the default group if you want.

**Comment by MilanMasek** [comment]
Thank you for explaining and linking to the issue where this has been solved and merged to 2.6.x and 3.x branches.

**Comment by MilanMasek** [comment]
Thank you for explaining and linking to the issue where this has been solved and merged to 2.6.x and 3.x branches.
